add-on/src/lib/ipfs-protocol.js

```diff
@@ -1,31 +1,57 @@
 const { mimeSniff } = require('./mime-sniff')
+const dirView = require('./dir-view')
+const PathUtils = require('ipfs/src/http/gateway/utils/path')
 
 exports.createIpfsUrlProtocolHandler = (getIpfs) => {
   return async (request, reply) => {
     console.time('[ipfs-companion] IpfsUrlProtocolHandler')
     console.log(`[ipfs-companion] handling ${request.url}`)
 
-    const path = request.url.split('ipfs://')[1]
+    let path = request.url.replace('ipfs://', '/')
+    path = path.startsWith('/ipfs') ? path : `/ipfs${path}`
+
     const ipfs = getIpfs()
 
     try {
-      const {data, mimeType} = await getDataAndGuessMimeType(ipfs, path)
-      console.log(`[ipfs-companion] returning ${path} as ${mimeType}`)
-      reply({mimeType, data})
+      const {data, mimeType, charset} = await getDataAndGuessMimeType(ipfs, path)
+      console.log(`[ipfs-companion] returning ${path} as mime ${mimeType} and charset ${charset}`)
+      reply({mimeType, data, charset})
     } catch (err) {
+      console.error('[ipfs-companion] failed to get data', err)
       reply({mimeType: 'text/html', data: `Error ${err.message}`})
     }
 
     console.timeEnd('[ipfs-companion] IpfsUrlProtocolHandler')
   }
 }
 
-function getDataAndGuessMimeType (ipfs, path) {
-  return new Promise((resolve, reject) => {
-    ipfs.files.cat(path, (err, res) => {
-      if (err) return reject(err)
-      const mimeType = mimeSniff(res, path)
-      resolve({mimeType, data: res.toString('utf8')})
-    })
-  })
+async function getDataAndGuessMimeType (ipfs, path) {
+  let data
+
+  try {
+    data = await ipfs.files.cat(path)
+  } catch (err) {
+    if (err.message.toLowerCase() === 'this dag node is a directory') {
```

Comment:
Ouch, is this the only way we can detect directories?

Reply:
I probably should have explained this a little...here goes:

It's similar to how the js-ipfs http gateway works. It tries to "resolve" a path (e.g. /ipfs/$hash/path/to/a/file) by using `ipfs.dag.get` to walk the path and eventually find the hash of the file at the end. It uses `Unixfs.unmarshal` on the end file to detect a directory and will [raise an error if so](https://github.com/ipfs/js-ipfs/blob/master/src/http/gateway/resolver.js#L94-L100) "this dag node is a directory". [If this error occurs](https://github.com/ipfs/js-ipfs/blob/master/src/http/gateway/resources/gateway.js#L39) then it'll try to get the links using `ipfs.object.get` and render a directory view.

Instead of walking the path using `ipfs.dag.get` (which isn't available in js-ipfs-api) we use `ipfs.files.cat` which has the advantage of being able to pass a path to it (so the walking is done for us) and is what we'd call to get file content anyway. It also returns the same error - "this dag node is a directory" for directories. We use `ipfs.ls` instead of `ipfs.object.get` because:

1. `ipfs.files.cat` resolved the hash but didn't return it to us so we can't use `ipfs.object.get` anyway since it needs a hash
2. We can pass a path to `ipfs.ls` and it'll give us the links

Is there a better way? @diasdavid